src/k8s/pkg/client/k8sd/query.go

```diff
@@ -6,14 +6,37 @@ import (
 	"strings"
 
 	apiv1 "github.com/canonical/k8s-snap-api/api/v1"
+	"github.com/canonical/k8s/pkg/log"
+	"github.com/canonical/k8s/pkg/utils/control"
 	"github.com/canonical/lxd/shared/api"
 )
 
-// query is a helper method to wrap common error checking and response handling.
-func query[T any](ctx context.Context, c *k8sd, method string, path string, in any, out *T) (T, error) {
-	if err := c.client.Query(ctx, method, apiv1.K8sdAPIVersion, api.NewURL().Path(strings.Split(path, "/")...), in, out); err != nil {
-		var zero T
-		return zero, fmt.Errorf("failed to %s /%s: %w", method, path, err)
+// query is a helper method for sending requests to the k8sd client with common error checking and automatic retries.
+// It retries on temporary microcluster errors and returns the deserialized response.
+func query[T any](ctx context.Context, c *k8sd, method, path string, in any, out *T) (T, error) {
+	var result T
+
```

Comment:
From what I've seen, we usually provide the structs where we unmarshal the response in `out`. I think this might be a minor nit, but it could be worthwhile to add a check to validate if the generic `out` pointer is nil and return early, instead of panicking down the road:
```suggestion
func query[T any](ctx context.Context, c *k8sd, method, path string, in any, out *T) (T, error) {
	var result T
	if out == nil {
		return result, fmt.Errorf("out must be a non-nil pointer")
	}
```